Add registration api

andreyatx opened the pull request:
Add registration functionality
Add logout feature

andreyatx committed:
refactor(SignUp): add dispatch

andreyatx committed:
feat(Menu): add logout handler

andreyatx committed:
Merge branch 'develop' into task/42-registration-api

andreyatx committed:
Merge branch 'develop' into task/42-registration-api

andreyatx committed:
feat(Home): add logout link

andreyatx committed:
test(Menu,Home): fix

andreyatx committed:
test(Home): fix provider issue

DmitriiNazimov commented:
При логауте надо чтобы делало редирект на главную и показывалось уведомление "Мы будем скучать!", а то сейчас как-будто ничего не происходит.

sdelfi commented on packages/client/src/pages/SignUp/SignUp.tsx:
```diff
@@ -1,23 +1,31 @@
+
+  const { error, isLoading, isAuthenticated } = useAppSelector(authSelectors.all);
+  if (isAuthenticated) {
+    navigate(Paths.Home);
```
Можно показывать toast об успешной регистрации.
Но механизм проверки на авторизацию для роутов еще поменяю (protected routes, авторизованный не может зайти на login, signup страницы, неавторизованный не может попасть в приватные разделы), пока можно, наверное, оставить так.

sdelfi commented on packages/client/src/pages/SignUp/SignUp.tsx:
```diff
@@ -37,6 +49,7 @@ export const SignUp: FC = () => {
         type="email"
```
Мы планировали использовтаь единый стиль для форм, т.к. настройки полей отдельно и потом в цикле выводить.
В signin и в user profile секции так сделано уже. 
Хотя это не registration api, но при желании можно отрефакторить.
С другой стороны - грядет еще интеграция вализации форм, возможно, там структура поменяется.

sdelfi commented on packages/client/src/pages/SignUp/SignUp.tsx:
```diff
@@ -85,11 +103,13 @@ export const SignUp: FC = () => {
       />
       <div className="form__buttons-wrapper">
-        <Button text="Зарегистрироваться" type="submit" variant={ButtonVariant.Primary} />
+        {error && `Error: ${error}`}
```
Уже интегрированы toast. Посмотри пример в signin компоненте

DmitriiNazimov commented:
Аналогично и при регистрации: 
1. Редирект на главную вроде есть, но мне показалось что не наверх страницы, а внизу страница остается. Если так, то надо принудительно скроллить наверх.
2. Нужно чтобы показывалось уведомление "Успех! Теперь одним танкистом больше!"

sdelfi commented on packages/client/src/components/Menu/Menu.tsx:
```diff
@@ -18,6 +21,10 @@ export const Menu: FC = () => {
   });
 
+  const logoutHandler = () => {
+    dispatch(authThunks.logout());
```
Я бы добавил async + await  (дождаться выполнения thunk) и после сделать редирект на главную.
А так все отлично в плане логики

sdelfi reviewed (approved):
По registration API замечаний нет, все отлично.

Есть пару моментов, оставил комментарии, но, возможно, это не в этом PR нужно править.

andreyatx committed:
refactor(SignUp): review changes

add toast notificatons
change input field

andreyatx committed:
refactor(client/src): review changes

* feat(client/src): add logout
* add Navigation component
* refactor Menu
* add notifications for sign-up and logout

DmitriiNazimov committed:
refactor(global): team live-coding

DmitriiNazimov commented:
Тут работы еще не закончены, PR не готов для полноценного ревью.

DmitriiNazimov commented on packages/client/src/pages/SignUp/SignUp.tsx:
```diff
@@ -1,95 +1,67 @@
 import { FormField } from '../../components/FormField';
 import { Paths } from '../../config/constants';
+import { authActions, authSelectors, authThunks, useAppDispatch, useAppSelector } from '../../store';
+import { formInitialState, signUpInputFields as signUpFieldList } from './data';
```
Тут импорт переименовался, а не сам signUpInputFields. Надо переименовать signUpInputFields в signUpFieldList

DmitriiNazimov reviewed (changes requested):
Надо выполнить все TODO из последнего пуша.

sapomaro commented on packages/client/src/layouts/Root/Root.tsx:
```diff
@@ -23,10 +23,12 @@ export const Root: FC<RootProps> = ({ children }) => {
       <header>
-        <Menu />
+        <BurgerMenu />
         {location?.pathname !== Paths.Game && (
```
Имеет смысл отходить от практики использования оператора && в сторону тернарного ? (https://developer.mozilla.org/ru/docs/Web/JavaScript/Reference/Operators/Conditional_Operator) - это позволит избежать появления при рендеринге случайных нулей и другого возможного мусора. См. раздел "Условный рендеринг" в https://habr.com/ru/post/541320/.

DmitriiNazimov replied:
Да, тернарный оператор лучше.

DmitriiNazimov replied:
Не стал это исправлять в одном месте здесь. По хорошему надо пройтись по всему проекту и исправить. Добавлю в задачи.

andreyatx committed:
refactor(client/src): team review

andreyatx committed:
refactor(store): fix naming

andreyatx committed:
refactor(client/src): fix lint

DmitriiNazimov committed:
refactor(global): close most of TODO's

DmitriiNazimov committed:
refactor(CSS): optimise CSS

DmitriiNazimov committed:
refactor(FormField): add TODO for next task

DmitriiNazimov commented:
@Olegas этот PR проверен командой и готов к ревью.

Olegas commented on packages/client/src/components/MenuLink/MenuLink.tsx:
```diff
@@ -6,14 +6,14 @@ import { NavLink } from 'react-router-dom';
 import { MenuLinkProps } from './typings';
 
-export const MenuLink: FC<MenuLinkProps> = ({ title, to, handleNavigate }) => {
+export const MenuLink: FC<MenuLinkProps> = ({ title, to, clickHandler }) => {
```
Обычно обработчики событий имеют названия вида onSomething

andreyatx replied:
Я так понимаю, название самой функции, где прописана логика можно оставить clickHandler? @Olegas 
А проп который висит на MenuLink назвать onClick (переименовал всё на всякий :) )

Olegas replied:
@andreyatx Название функции может быть какое угодно, а проп лучше называть onClick.

Olegas commented on packages/client/src/components/Navigation/Navigation.tsx:
```diff
@@ -0,0 +1,45 @@
+    if (linkName === 'logout') {
+      await dispatch(authThunks.logout());
+
+      if (!error) {
```
Вот это очень странный участок. Error тут интересует на момент старта обработчика? Или расчет на то, что error поменяется в процессе await dispatch?

sdelfi replied:
Во время  await dispatch в error может прийти ошибка с API.
Но наверное да, error нужно слушать в корне компонента в данном случае.
Мы обсуждали вывод ошибки с командой вчера, но упустили этот момент.

DmitriiNazimov replied:
Станислав, поправишь этот момент? В этом PR.

DmitriiNazimov replied:
@Olegas мы посовещались и пришли к выводу, что в данном случае изменения кода не требуются. 

Код работает как Вы и предположили - error может поменяться в процессе await dispatch. 

Возможно по этому поводу мог бы пригодится комментарий, но мы проговорили этот момент на синке, все поняли как это работает и поэтому комментарий не требуется. 

Т.е. если замечание было только по поводу того, что этот участок кода не очень понятен, то этот вопрос мы в команде решили, а если тут есть какие-то ошибки которые мы не видим, то просьба указать на них конкретно :)

Olegas replied:
Я считаю что этот участок кода в принципе не будет работать правильно. Я напишу пример. Пока не сливайте. С меня объяснение почему.

Olegas replied:
Написал ниже пояснение с примером - почему это не работает

Olegas commented on packages/client/src/store/features/auth/authSlice.ts:
```diff
@@ -34,6 +41,8 @@ export const authSlice = createSlice({
         state.error = '';
       })
+      // Здесь нет fulfilled т.к. вызывается thunk Me и уже там отрабатывает fulfilled,
+      // который ставит isLoading=false
```
Полезные комментарии - 👍

Olegas commented on packages/client/src/store/features/ui/typings.ts:
```diff
@@ -0,0 +1,3 @@
+export interface uiState {
```
Типы и интерфейсы обычно именуются с большой буквы. А UI это к тому же еще и аббревиатура

andreyatx committed:
refactor(client/src): review changes

andreyatx committed:
Merge branch 'task/42-registration-api' of https://github.com/FrontHeads/tanchiki into task/42-registration-api

sdelfi committed:
feat(router): add protected and public routes. Change tests

* fix(router): add protected and public routes. Change tests

add protected routes
add test setup script with global mocks
change redux store structure

BREAKING CHANGE:

* fix(router): fix router test issues

* refactor(jest): change testing-utils

* fix(test): fix tests

* fix(test): fix tests

* fix: review issues correction

* refactor(AuthSlice): add new selector

* refactor: change comment format

* refactor(PublicRoutes): add PublicRoute notification

* fix(authSlice): remove isAuthenticated property from auth store

* fix(jest): fix Logo test issue

andreyatx committed:
refactor(MenuLink): minor fix

sdelfi committed:
fix: fix authSlice typings

sdelfi committed:
Merge branch 'develop' into task/42-registration-api

andreyatx committed:
style(index.css): fix anchor tag style

DmitriiNazimov reviewed (changes requested):
Изменения которые я делал затерлись следующими коммитами. Необходимо их вернуть.

andreyatx committed:
refactor(client/src): restore previous commits

DmitriiNazimov committed:
fix(global): minor fix

DmitriiNazimov commented:
Кажется все замечания исправлены.

DmitriiNazimov committed:
chore(eslint): use type alias instead interface

Olegas reviewed (changes requested):
Участок с обработкой ошибки неработоспособен. С меня пример.

Olegas commented:
Вот пример: https://playcode.io/1027863

Смотрите что происходит... Компонент - это функция. Каждый раз при запуске этой функции объявляются и инициализируются все переменные. Одна из таких переменных - clickHandler в которой лежит анонимная асинхронная функция. Эта функция создает замыкание через которое видит переменную error содержащую текущую ошибку, полученную из redux store. Она пока пустая. Пока все хорошо

Далее, пользователь нажимает и срабатывает clickHandler. Он запускает API-вызов и ожидает. Где-то внутри, во время этого API-вызова происходит ошибка. Она записывается в redux store после чего у нашего компонента случается перерисовка, вызванная изменениями в store.

Снова вызывается функция компонента, инициализируются переменные, одна их которых содержит анонимную асинхронную функцию, которая создает замыкание и видит переменную error которая содержит ошибку. ВНИМАНИЕ - в clickHandler лежит другая, новая функция, никак не связанная с тем, что мы получили в первом абзаце.

Далее заканчивается await и управление возвращается в наш первый clickHandler, в ту точку где мы делали await, он смотрит на переменную error - она пустая. Потому что это другое лексическое окружение, другое замыкание, в этой "параллельной вселенной" error не заполнен. Вот тут все и сломалось.

В моем примере по ссылке функция longFunction имитирует API-вызов которого мы дожидаемся. Внутри timeout имитирует поход в сеть. Т.к. у меня там нет redux я просто передал в нее сеттер что бы сымитировать выставление ошибки. Как мы видим, компонент перерисовался, ошибка на экран выведена. Но в логе консоли видно что после await ошибка все еще пустая.

sdelfi committed:
refactor(Logout): change logout onclick event handler

sdelfi commented:
> Вот пример: https://playcode.io/1027863

Спасибо большое, что пояснили. Полезно знать. 
В итоге переделал немного код, теперь работает как нужно, ошибки отлавилваются.. 
Thunk позволяет через  then().catch() получить результат выполнения и отловить ошибки.

Olegas commented:
Теперь все супер, молодцы!
